src/Pate/AssumptionSet.hs

```diff
@@ -0,0 +1,340 @@
+fromExprBindings ::
+  forall sym.
+  W4.IsSymExprBuilder sym =>
+  OrdF (W4.SymExpr sym) =>
+  ExprBindings sym ->
+  AssumptionSet sym
+fromExprBindings binds = AssumptionSet { asmPreds = mempty, asmBinds = MapF.map SetF.singleton binds }
+
+
+fromPred ::
+  forall sym.
+  W4.IsSymExprBuilder sym =>
+  W4.Pred sym ->
+  AssumptionSet sym
+fromPred p = AssumptionSet (SetF.singleton p) MapF.empty
+
+exprBinding ::
+  forall sym tp.
+  W4.IsSymExprBuilder sym =>
+  -- | source expression
+  W4.SymExpr sym tp ->
+  -- | target expression
+  W4.SymExpr sym tp ->
+  AssumptionSet sym
+exprBinding eSrc eTgt = case testEquality eSrc eTgt of
+  Just Refl -> mempty
+  _ -> mempty { asmBinds = (MapF.singleton eSrc (SetF.singleton eTgt)) }
+
+natBinding ::
+  forall sym.
+  W4.IsSymExprBuilder sym =>
+  -- | source expression
+  W4.SymNat sym ->
+  -- | target expression
+  W4.SymNat sym ->
+  AssumptionSet sym
+natBinding n1 n2 = exprBinding (W4.natToIntegerPure n1) (W4.natToIntegerPure n2)
+
+ptrBinding ::
+  W4.IsSymExprBuilder sym =>
+  CLM.LLVMPtr sym w ->
+  CLM.LLVMPtr sym w ->
+  AssumptionSet sym
+ptrBinding (CLM.LLVMPointer reg1 off1) (CLM.LLVMPointer reg2 off2) =
+  let
+    regBind = natBinding reg1 reg2
+    offBind = exprBinding off1 off2
+  in (regBind <> offBind)
+
+
+
+macawRegBinding ::
+  W4.IsSymExprBuilder sym =>
+  MS.ToCrucibleType tp ~ MS.ToCrucibleType tp' =>
+  sym ->
+  -- | value to rebind
+  PSR.MacawRegEntry sym tp ->
+  -- | new value
+  PSR.MacawRegEntry sym tp' ->
+  AssumptionSet sym
+macawRegBinding _sym var val = do
+  case PSR.macawRegRepr var of
+    CLM.LLVMPointerRepr _ -> ptrBinding (PSR.macawRegValue var) (PSR.macawRegValue val)
+    CT.BoolRepr -> exprBinding (PSR.macawRegValue var) (PSR.macawRegValue val)
+    CT.StructRepr Ctx.Empty -> mempty
+    _ -> panic Rewriter "macawRegBinding" ["Unexpected macaw types"
+                                          , show (PSR.macawRegRepr var)
+                                          , show (PSR.macawRegRepr val)
+                                          ]
+
+
+
+-- | Rewrite the given 'f' with any bindings in the given 'AssumptionSet'.
+--   Bindings are applied repeatedly to each component expression until a fixpoint
+--   is reached or a loop is detected.
+apply ::
+  forall sym f m t solver fs.
+  PEM.ExprMappable sym f =>
+  IO.MonadIO m =>
+  sym ~ (W4B.ExprBuilder t solver fs) =>
+  sym ->
+  AssumptionSet sym ->
+  f ->
+  m f
+apply sym asm f = do
+  cache <- IO.liftIO WEH.freshVarBindCache
+  apply' sym cache asm f
+
+-- | Rewrite the given 'f' with any bindings in the given 'AssumptionSet'.
+--   Bindings are applied repeatedly to each component expression until a fixpoint
+--   is reached or a loop is detected.
+apply' ::
+  forall sym f m t solver fs.
+  PEM.ExprMappable sym f =>
+  IO.MonadIO m =>
+  sym ~ (W4B.ExprBuilder t solver fs) =>
+  sym ->
+  VarBindCache sym ->
+  AssumptionSet sym ->
+  f ->
+  m f
+apply' sym cache asm f = do
+  let
```

Comment:
If this does need to be exported, a more descriptive name like `applyWithCache` would be better than `apply'`---understanding what the prime means in isolation is difficult